units/bson_to_parquet_backfill.go

```diff
@@ -0,0 +1,184 @@
+	job.Base `bson:"metadata" json:"metadata" yaml:"metadata"`
+
+	env         cedar.Environment
+	queue       amboy.Queue
+	controller  *model.BatchJobController
+	migrationID string
+}
+
+func init() {
+	registry.AddJobType(bsonToParquetBackfillJobName, func() amboy.Job { return makeBSONToParquetBackfillJob() })
+}
+
+func makeBSONToParquetBackfillJob() *bsonToParquetBackfillJob {
+	return &bsonToParquetBackfillJob{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    bsonToParquetBackfillJobName,
+				Version: 1,
+			},
+		},
+	}
+}
+
+// NewBSONToParquetBackfillJob returns a new Amboy job that does batched
+// migrations of existing BSON test results to Apache Parquet test results
+// stored in the Presto bucket.
+func NewBSONToParquetBackfillJob(id string) amboy.Job {
+	j := makeBSONToParquetBackfillJob()
+	j.SetID(fmt.Sprintf("%s.%s", bsonToParquetBackfillJobName, id))
+
+	return j
+}
+
+func (j *bsonToParquetBackfillJob) Run(ctx context.Context) {
+	defer j.MarkComplete()
+
+	if j.env == nil {
+		j.env = cedar.GetEnvironment()
+	}
+	if j.queue == nil {
+		j.queue = j.env.GetRemoteQueue()
+	}
+
+	controller, err := model.FindBatchJobController(ctx, j.env, bsonToParquetBackfillJobName)
+	if err != nil {
+		j.AddError(errors.Wrap(err, "getting batch job controller"))
+		return
+	}
+	if controller.Iterations <= 0 {
+		controller.Iterations = 1
+	}
+	if controller.Timeout <= 0 {
+		controller.Timeout = time.Minute
+	}
+	j.controller = controller
+
+	var cancel context.CancelFunc
+	ctx, cancel = context.WithTimeout(ctx, j.controller.Timeout)
+	defer cancel()
+
+	for i := 0; i < j.controller.Iterations; i++ {
+		count, err := j.runIteration(ctx)
+		if err != nil {
+			j.AddError(err)
+			return
+		}
+		if count == 0 {
+			return
+		}
+	}
+}
+
+func (j *bsonToParquetBackfillJob) runIteration(ctx context.Context) (int64, error) {
+	var count int64
+	start := time.Now().UTC()
+	collection := j.env.GetDB().Collection(j.controller.Collection)
+
+	// This is really ugly, but unfortunately Mongo does not allow setting
+	// limits on UpdateMany operations so we have to do a BulkWrite with
+	// the batch size number of UpdateOne operations.
+	updates := make([]mongo.WriteModel, j.controller.BatchSize)
+	for i := 0; i < j.controller.BatchSize; i++ {
+		updates[i] = mongo.NewUpdateOneModel().
+			SetFilter(bson.M{"$or": []bson.M{
+				{model.TestResultsMigrationKey: bson.M{"$exists": false}},
+				{bsonutil.GetDottedKeyName(model.TestResultsMigrationKey, model.MigrationStatsVersionKey): bson.M{"$lt": j.controller.Version}},
+				{"$and": []bson.M{
+					{bsonutil.GetDottedKeyName(model.TestResultsMigrationKey, model.MigrationStatsStartedAtKey): bson.M{"$lte": time.Now().Add(-j.controller.Timeout)}},
+					{bsonutil.GetDottedKeyName(model.TestResultsMigrationKey, model.MigrationStatsCompletedAtKey): nil},
+				}},
+			}}).
+			SetUpdate(bson.M{"$set": bson.M{
+				model.TestResultsMigrationKey: &model.MigrationStats{
+					MigratorID: j.ID(),
```

Comment:
Is this migration ID/started at staleness check for retrying the backfill when it gets interrupted? If so, can it be made a retryable job and set the `WaitUntil` time instead? Seems like it would cut down on some of the querying complexity and bookkeeping on the start time.

Reply:
It is also so we can run these jobs in parallel, so there may be multiple migrator IDs doing work at the same time and we need to be able to distinguish when fetching the docs in the next step, and updating them as completed in the last step.